File: BackTracking/JS/PowerSet.js
init();

function init(){
    
    let arr = [0,1,2];
    let l = arr.length;
    let result = new Array();
    for(let i = 0 ; i < Math.pow(2,l)-1; i++){
       let curr = new Array();
       for(let j = 0 ; j < l; j++){
           if(i & (1<<j)) curr.push(arr[j]);
       }
       result.push(curr);
    }
    console.log(...result);

}

File: BackTracking/CPP/PowerSet.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> powerSet(vector<int> &arr, vector<int> temp, vector<vector<int>> &result){
    int n = arr.size();
   //	The expression i & (1 << j) checks whether the j-th bit in the binary representation of i is 1. 
    for(auto i = 0 ; i < pow(2,n) - 1 ; i++){
        temp.clear();
        for(auto j = 0 ; j < n ; j++){
            if(i & (1<<j)) temp.push_back(arr[j]);
        }
        result.push_back(temp);
    }
    return result;
}


int main(){

   vector<int> arr = {0,1,2};
   vector<int> temp;
   vector<vector<int>> result;
   powerSet(arr,temp,result);
   
   for(auto k : result){
    for(auto x : k){
        cout<<x<<" ";
    }
    cout<<endl;
   }

}